File: src/main/java/uk/co/eelpieconsulting/feedlistener/rss/images/BodyHtmlImageExtractor.java
package uk.co.eelpieconsulting.feedlistener.rss.images;

import com.google.common.base.Strings;
import com.rometools.rome.feed.synd.SyndEntry;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.htmlparser.NodeFilter;
import org.htmlparser.Parser;
import org.htmlparser.Tag;
import org.htmlparser.filters.TagNameFilter;
import org.htmlparser.util.NodeList;
import org.htmlparser.util.ParserException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import uk.co.eelpieconsulting.feedlistener.rss.RssFeedItemBodyExtractor;

@Component
public class BodyHtmlImageExtractor {

    private static Logger log = LogManager.getLogger(BodyHtmlImageExtractor.class);

    private final RssFeedItemBodyExtractor rssFeedItemBodyExtractor;

    @Autowired
    public BodyHtmlImageExtractor(RssFeedItemBodyExtractor rssFeedItemBodyExtractor) {
        this.rssFeedItemBodyExtractor = rssFeedItemBodyExtractor;
    }

    public String extractImageFrom(SyndEntry item) {
        final String itemBody = rssFeedItemBodyExtractor.extractBody(item);
        if (!Strings.isNullOrEmpty(itemBody)) {
            String imageSrc = extractImage(itemBody);
            if (!Strings.isNullOrEmpty(imageSrc)) {
                return imageSrc;
            }
        }
        return null;
    }

    private String extractImage(final String itemBody) {
        try {
            final Parser parser = new Parser();
            parser.setInputHTML(itemBody);
            NodeFilter tagNameFilter = new TagNameFilter("img");
            NodeList imageNodes = parser.extractAllNodesThatMatch(tagNameFilter);
            log.debug("Found images: " + imageNodes.size());
            return extractFirstImage(imageNodes);

        } catch (ParserException e) {
            log.warn("Failed to parse item body for images", e);
        }
        return null;
    }

    private String extractFirstImage(NodeList imageNodes) {
        if (imageNodes.size() == 0) {
            return null;
        }

        final Tag imageTag = (Tag) imageNodes.elementAt(0);
        final String imageSrc = imageTag.getAttribute("src");
        log.debug("Found first image: " + imageTag.toHtml() + ", " + imageSrc);
        return imageSrc;
    }

}

File: src/main/kotlin/uk/co/eelpieconsulting/feedlistener/rss/images/BodyHtmlImageExtractor.kt
package uk.co.eelpieconsulting.feedlistener.rss.images

import com.google.common.base.Strings
import com.rometools.rome.feed.synd.SyndEntry
import org.apache.logging.log4j.LogManager
import org.htmlparser.NodeFilter
import org.htmlparser.Parser
import org.htmlparser.Tag
import org.htmlparser.filters.TagNameFilter
import org.htmlparser.util.NodeList
import org.htmlparser.util.ParserException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import uk.co.eelpieconsulting.feedlistener.rss.RssFeedItemBodyExtractor

@Component
class BodyHtmlImageExtractor @Autowired constructor(private val rssFeedItemBodyExtractor: RssFeedItemBodyExtractor) {

    private val log = LogManager.getLogger(BodyHtmlImageExtractor::class.java)

    fun extractImageFrom(item: SyndEntry): String? {
        val itemBody = rssFeedItemBodyExtractor.extractBody(item)
        return if (itemBody != null &&!Strings.isNullOrEmpty(itemBody)) {
            val imageSrc = extractImage(itemBody)
            if (!Strings.isNullOrEmpty(imageSrc)) imageSrc else null
        } else {
            null
        }
    }

    private fun extractImage(itemBody: String): String? {
        return try {
            val parser = Parser()
            parser.setInputHTML(itemBody)
            val tagNameFilter: NodeFilter = TagNameFilter("img")
            val imageNodes: NodeList = parser.extractAllNodesThatMatch(tagNameFilter)
            log.debug("Found images: " + imageNodes.size())
            extractFirstImage(imageNodes)
        } catch (e: ParserException) {
            log.warn("Failed to parse item body for images", e)
            null
        }
    }

    private fun extractFirstImage(imageNodes: NodeList): String? {
        if (imageNodes.size() == 0) {
            return null
        }
        val imageTag = imageNodes.elementAt(0) as Tag
        val imageSrc = imageTag.getAttribute("src")
        log.debug("Found first image: " + imageTag.toHtml() + ", " + imageSrc)
        return imageSrc
    }

}
